### PR TITLE
Client: LocalBackend

### LINKED ISSUE #9: Client: Local backend
Implement `LocalBackend` which will store trials on local disk in form of files.

Epic https://github.com/IceKhan13/purplecaffeine/issues/1

### LINKED ISSUE #22: Client: Trial save & read backend
### What is the expected enhancement?
Implement the save and read backend of Trial class.

### LINKED ISSUE #24: Trial: add tags
### What is the expected enhancement?

Add tags to Trial class.

### PR DESCRIPTION
**Pull request by mickahell**
### Summary
Create the localBackend class.

#### Backend
- [x] Save trial into localbackend
- [x] Read trial from localbackend
- [x] Unit tests of backend

#### Trial
- [x] Save trial of into any Backend
- [x] Read trial of from any Backend
- [x] Transform arrays
from `arrays[array: Union[np.ndarray, List[Any]]]`
into `arrays[name: str, array: Union[np.ndarray, List[Any]]]`
- [x] Add tags as `tags[tag: str]`
- [x] Unit tests

#### Serializing Trial
- [x] TrialEncoder
- [x] TrialDecoder
- [x] Integrate everything together
- [x] Unit tests

---
Closes #9 
Closes #22 
Closes #24

### CODE REVIEW

**Comment by mickahell** [comment]
@IceKhan13 do you know if Qiskit has some built-in method to :
- convert circuits to something usable (& vice versa) ?
- convert backends to something usable (& vice versa) ?
- operators to something usable (& vice versa) ?
- result to something usable (& vice versa) ?

**Comment by IceKhan13** [comment]
@mickahell here you go https://github.com/Qiskit/qiskit-ibm-runtime/blob/main/qiskit_ibm_runtime/utils/json.py#L201 :)

**Comment by mickahell** [comment]
@IceKhan13 ready for your review :)

File: client/purplecaffeine/backend.py [hunk 1]
```diff
@@ -1,14 +1,16 @@
 """Backend."""
+import json
 
 
 class BaseBackend:
     """Base backend class."""
 
-    def save_trial(self, trial):
+    def save_trial(self, name: str, trial_json: json):
```
Comment (IceKhan13) [inline client/purplecaffeine/backend.py:8]: Let's keep type `Trial` as later on we can change serialization from json to more performant binary format :)

Comment (IceKhan13) [inline client/purplecaffeine/backend.py]: Let's keep type `Trial` as later on we can change serialization from json to more performant binary format :)

File: client/purplecaffeine/backend.py [hunk 2]
```diff
@@ -20,14 +22,40 @@ def __init__(self, path: str):
         """Local backend.
 
         Args:
-            path: folder where trail data will be saved.
+            path: folder where trial data will be saved.
         """
         self.path = path
 
-    def save_trial(self, trial):
+    def save_trial(self, name: str, trial_json: json) -> str:
```
Comment (IceKhan13) [inline client/purplecaffeine/backend.py]: same here

File: client/purplecaffeine/backend.py [hunk 3]
```diff
@@ -20,14 +22,40 @@ def __init__(self, path: str):
...
-            trial: trial to save
+            name: name of the trial
+            trial_json: encode trial to save
+
+        Returns:
+            self.path: path of the trial file
         """
+
+        with open(
+            self.path + "/" + name + ".json", "w", encoding="utf-8"
```
Comment (IceKhan13) [inline client/purplecaffeine/backend.py]: paths should be formated using `os.path.join("bla-bla", "bla-bla")` :)

File: client/purplecaffeine/backend.py [hunk 4]
```diff
@@ -20,14 +22,40 @@ def __init__(self, path: str):
...
+        """Read a given trial file.
+
+        Args:
+            name: name of the trial
+
+        Returns:
+            trial_json: Json object of a trial
+        """
+        with open(
+            self.path + "/" + name + ".json", "r", encoding="utf-8"
```
Comment (IceKhan13) [inline client/purplecaffeine/backend.py]: same as above

File: client/purplecaffeine/backend.py [hunk 5]
```diff
@@ -20,14 +22,40 @@ def __init__(self, path: str):
...
+            trial_json: encode trial to save
+
+        Returns:
+            self.path: path of the trial file
         """
+
+        with open(
+            self.path + "/" + name + ".json", "w", encoding="utf-8"
+        ) as trial_file:
+            trial_file.write(trial_json)
```
Comment (IceKhan13) [inline client/purplecaffeine/backend.py]: Maybe we should use `json.dump` with `JsonDecoder` for Trial. What do you think? Something like

```python
class OurCustomEncoderForTrial(JsonEncoder):
     """Encoder which serialize content of Trial into json/."""

with open() as f:
    json.dump(f, trial, cls=OurCustomEncoderForTrial)
```

File: client/purplecaffeine/trial.py [hunk 6]
```diff
@@ -93,20 +99,20 @@ def add_operator(self, name: str, operator: Operator):
...
         """Adds artifacts path to trial data.
 
         Args:
             name: name of the file
-            path_to_file: path to access to the file
+            artifact: file object
         """
-        if os.stat(path_to_file).st_size >= Configuration.MAX_SIZE:
+        if asizeof.asizeof(artifact) >= Configuration.MAX_SIZE:
             print(
```
Comment (IceKhan13) [inline client/purplecaffeine/trial.py]: let's convert this to `logging.warninig`

File: client/purplecaffeine/trial.py [hunk 7]
```diff
@@ -117,13 +123,92 @@ def add_text(self, title: str, text: str):
...
+        """Adds any tag to trial data.
+
+        Args:
+            tag: word of your tag
+        """
+        self.tags.append(tag)
+
+    def save_trial(self):
+        """Serialize Trial into Json format and save into Backend."""
+        circuits_encoder = []
```
Comment (IceKhan13) [inline client/purplecaffeine/trial.py]: If we will have our own TrialEncoder and Decoder then all this code will go away there and Trial class will be super simple and clean :)

**Review by IceKhan13 — CHANGES_REQUESTED** [review]
Some coding style changes and maybe file rearranging is up to you.

Overall looks very good! We are moving to right direction! Love the tests :) 

Thank you!

File: client/purplecaffeine/backend.py (hunk 5 above)
Comment (mickahell) [reply client/purplecaffeine/backend.py]: It was what I was thinking at first but as the style object of Qiskit keep changing overtime I wouldn't want for us to have to often deprecated format. Furthermore, we indeed style have tuple object...
Let's try something, you'll tell me :)

**Comment by mickahell** [comment]
@IceKhan13 done :)
I created our own Encoder/Decoder and I tried to do it as much normalize as possible.

**Review by IceKhan13 — APPROVED** [review]
So clean now!

**Pull request merged**